### ISSUE TITLE
symfony/http-client requests always blocking untill first byte

### ISSUE THREAD

**Issue by Bonemeijer** [symfony/symfony#35746]
**Symfony version(s) affected**: v5.0.4

**Description**  
We're using `symfony/http-client` to make parallel requests to remote API's. I have noticed that when it takes a while to connect to the remote server, the requests are actually blocking.

Only after either headers or first byte is received from the remote, then the request starts to become asynchronous.

**How to reproduce**  

Make the following `sleep.php` available via a local HTTP server

```
<?php

echo mt_rand().PHP_EOL;

// without flushing the buffer, the requests are blocking
//while (ob_get_contents() !== false) {
//    ob_get_flush();
//    flush();
//}

sleep(5);

echo mt_rand();
```

Then use the following script for testing

```
<?php
// after running `composer require symfony/http-client`

use Symfony\Component\HttpClient\CurlHttpClient;

include __DIR__ . '/vendor/autoload.php';

for ($i=0; $i<3; $i++) {
    $client = new CurlHttpClient();
    $client->request('GET', 'https://.../sleep.php?'.mt_rand());
}
```

**Additional context**
When I use curl directly, the requests *are* correctly asynchronous, wether or not the output buffer is flushed early.

For example, the following script correcly takes 5 seconds:

```
<?php

$multiCurl = [];
$result    = [];
$mh        = curl_multi_init();

for ($i = 0; $i < 3; $i++) {
    $multiCurl[$i] = curl_init();
    curl_setopt($multiCurl[$i], CURLOPT_URL, 'https://.../sleep.php');
    curl_setopt($multiCurl[$i], CURLOPT_HEADER, 0);
    curl_setopt($multiCurl[$i], CURLOPT_RETURNTRANSFER, 1);
    curl_multi_add_handle($mh, $multiCurl[$i]);
}

$stillRunning = null;
do {
    curl_multi_exec($mh, $stillRunning);
} while ($stillRunning > 0);

foreach ($multiCurl as $k => $ch) {
    $result[$k] = curl_multi_getcontent($ch);
    curl_multi_remove_handle($mh, $ch);
}
curl_multi_close($mh);

print_r($result);
```

**Comment by nicolas-grekas** [comment]
Hello, thanks for the report.
We just added a few sentences about this behavior in the doc, you might want to check https://symfony.com/doc/current/components/http_client.html#handling-exceptions, especially around the sentence
> https://symfony.com/doc/current/components/http_client.html#handling-exceptions

TLDR: response are async *unless* you don't keep a reference to them (you don't in the for loop in your example).

**Comment by Bonemeijer** [comment]
Thank you for your quick response.

I'm not keeping the reference in both the case where it does work as in the case where it doesn't.

I have just retested while keeping the reference, and the issue still persists. I have updated the testcase to include keeping of the reference.

**Comment by nicolas-grekas** [comment]
How do you run the local PHP server?

**Comment by Bonemeijer** [comment]
I'm using;
- MOD PHP 7.4.2 
   `php.x86_64                              7.4.2-1.el7.remi           @remi-php74`
- Apache/2.4.41 (Unix)
- vagrant box with CentOS Linux release 7.7.1908 (Core)

**Comment by nicolas-grekas** [comment]
Oh, OK, here is why: you create a new client in each iteration of the loop.
Try this, it will last 5s and not 15s.
```php
$responses = [];
$client = new CurlHttpClient();

for ($i=0; $i<3; $i++) {
    $responses[] = $client->request('GET', 'https://httpbin.org/delay/5');
}
```

**Comment by Bonemeijer** [comment]
Thanks. Indeed, when I test it like that, the requests are asynchronous again.

But is the sending of asynchronous requests only supported when we use the same client?

In our case, we want to send multiple requests to different remote API's. Which each have their own specific configuration in the `framework.http_client` config section.

Please consider the next example, where I'm using different clients:

```
<?php
// after running `composer require symfony/http-client`

use Symfony\Component\HttpClient\CurlHttpClient;
use Symfony\Contracts\HttpClient\ResponseInterface;

include __DIR__ . '/vendor/autoload.php';

$clientOne   = new CurlHttpClient();
$clientTwo   = new CurlHttpClient();
$clientThree = new CurlHttpClient();

/** @var ResponseInterface[] $responses */
$responses = [
    $clientOne->request('GET', 'https://.../sleep.php?' . mt_rand()),
    $clientTwo->request('GET', 'https://.../sleep.php?' . mt_rand()),
    $clientThree->request('GET', 'https://.../sleep.php?' . mt_rand()),
];

$results = [];

foreach ($responses as $response) {
    $results[] = $response->getContent();
}

print_r($results);
```

**Comment by nicolas-grekas** [comment]
> But is the sending of asynchronous requests only supported when we use the same client?

yes absolutely, exactly like there is only one curl-multi handle in your example with raw curl.

> we want to send multiple requests to different remote API

you can do so with the same client.

> each have their own specific configuration in the framework.http_client config section

that's perfect. Under the hood, each configured client will be a wrapper on top of a single HTTP client that will allow making everything in parallel.

Check https://symfony.com/doc/current/components/http_client.html#scoping-client
which is the implem used.

**Comment by Bonemeijer** [comment]
Thank you very much for you patience and clear responses.

I did some additional testing in the framework and indeed it seems to be working as you say. 
I'm not sure why it didn't work before, but I'm guessing that must have had something to do with not keeping the reference to the response.
